File: src/tweeter.py
import json
import logging
import subprocess
import uuid
from datetime import datetime, timezone

import requests

from src.db import Database
from src.poster import Poster
from src.telegram_bot import TelegramReviewBot

logger = logging.getLogger(__name__)

TWEET_PROMPT = """You are the social media voice of a professional AI agent analytics platform.
Write ONE tweet about on-chain AI agent ecosystem stats.

Rules:
- Under 280 characters
- English only
- Crypto-native casual tone
- Include specific numbers from the data
- No links, no hashtags, no emojis
- Make it insightful, not just numbers
- Never mention any product or service name

Data:
{stats_json}

Write the tweet."""


class StatsCollector:
    def __init__(self, api_base_url: str):
        self.api_base_url = api_base_url.rstrip("/")

    def fetch_ecosystem_stats(self) -> dict | None:
        try:
            resp = requests.get(f"{self.api_base_url}/api/agents/filter-options", timeout=15)
            resp.raise_for_status()
            return resp.json()
        except Exception as e:
            logger.error("Failed to fetch ecosystem stats: %s", e)
            return None

    def fetch_trending_agents(self) -> list | None:
        try:
            resp = requests.get(f"{self.api_base_url}/api/agents/suggestions", timeout=15)
            resp.raise_for_status()
            data = resp.json()
            return data if isinstance(data, list) else data.get("data", [])
        except Exception as e:
            logger.error("Failed to fetch trending agents: %s", e)
            return None

    def fetch_top_agents(self) -> list | None:
        try:
            resp = requests.get(f"{self.api_base_url}/api/agents", params={"sort": "score", "pageSize": "5"}, timeout=15)
            resp.raise_for_status()
            data = resp.json()
            return data if isinstance(data, list) else data.get("data", [])
        except Exception as e:
            logger.error("Failed to fetch top agents: %s", e)
            return None

    def collect_all(self) -> dict:
        return {
            "ecosystem": self.fetch_ecosystem_stats(),
            "trending": self.fetch_trending_agents(),
            "top_agents": self.fetch_top_agents(),
        }


class Tweeter:
    DAILY_TWEET_CAP = 2

    def __init__(self, stats_collector: StatsCollector, poster: Poster, telegram: TelegramReviewBot, db: Database):
        self.stats = stats_collector
        self.poster = poster
        self.telegram = telegram
        self.db = db

    def _get_today_tweet_count(self) -> int:
        today = datetime.now(timezone.utc).strftime("%Y-%m-%dT00:00:00+00:00")
        result = (
            self.db.client.table("replies")
            .select("id")
            .eq("source_type", "original_tweet")
            .eq("status", "posted")
            .gte("posted_at", today)
            .execute()
        )
        return len(result.data)

    def generate_tweet(self, stats: dict) -> str | None:
        prompt = TWEET_PROMPT.format(stats_json=json.dumps(stats, indent=2))
        result = subprocess.run(
            ["claude", "-p", prompt, "--model", "sonnet"],
            capture_output=True, text=True,
        )
        text = result.stdout.strip()
        if not text or len(text) > 280:
            if text and len(text) > 280:
                text = text[:277] + "..."
        return text if text else None

    def run(self) -> dict:
        result = {"tweets_generated": 0}

        if self._get_today_tweet_count() >= self.DAILY_TWEET_CAP:
            result["skipped_reason"] = "daily_tweet_cap_reached"
            return result

        stats = self.stats.collect_all()
        if not any(stats.values()):
            result["skipped_reason"] = "no_stats_available"
            return result

        tweet_text = self.generate_tweet(stats)
        if not tweet_text:
            result["skipped_reason"] = "generation_failed"
            return result

        tweet_id = str(uuid.uuid4())
        self.db.save_reply({
            "id": tweet_id,
            "tweet_id": tweet_id,
            "draft_text": tweet_text,
            "final_text": None,
            "status": "pending",
            "source_type": "original_tweet",
            "reviewed_at": None,
            "posted_at": None,
            "created_at": datetime.now(timezone.utc).isoformat(),
        })

        import asyncio
        asyncio.run(self.telegram.send_review(
            {"content": "[Original Tweet]", "author_username": "agent_lynx"},
            tweet_text, tweet_id,
        ))

        result["tweets_generated"] = 1
        return result

File: src/tweeter.ts
import { Anthropic } from '@anthropic-ai/sdk';
import { v4 as uuidv4 } from 'uuid';
import { db } from './db';
import { telegram } from './telegram';
import { config } from './config';

const CHAIN_NAMES: Record<number, string> = {
  1: 'Ethereum',
  56: 'BNB Chain',
  143: 'Monad',
  8453: 'Base',
  42220: 'Celo',
};

const CHAIN_SLUGS: Record<number, string> = {
  1: 'ethereum',
  56: 'bsc',
  8453: 'base',
  42220: 'celo',
  143: 'monad',
};

const TWEET_TYPES = ['agent_highlight', 'anomaly', 'comparison'];

const SYSTEM_PROMPT = `You are a crypto-native analyst. You write extremely short, punchy tweets.
HARD LIMIT: Your output must NEVER exceed 160 characters total. No exceptions.
Style: casual, no hashtags, no URLs, no emojis, no product names.`;

const PROMPTS: Record<string, string> = {
  agent_highlight: `Here is agent data: {data_json}

Write ONE tweet about this agent's on-chain performance. Pick ONE interesting number (volume, P&L, or tx count) and make a punchy observation.

HARD LIMIT: 160 characters maximum. Output ONLY the tweet text, nothing else.`,

  anomaly: `Here is agent data: {data_json}

Write ONE tweet about an unusual pattern you see. Highlight ONE weird finding with a number. Contrarian tone.

HARD LIMIT: 160 characters maximum. Output ONLY the tweet text, nothing else.`,

  comparison: `Here is agent data: {data_json}

Write ONE tweet comparing agent activity across chains. Pick ONE cross-chain stat and make it punchy.

HARD LIMIT: 160 characters maximum. Output ONLY the tweet text, nothing else.`,
};

interface Agent {
  chain_id?: number;
  agent_id?: string;
  chainId?: number;
  agentId?: string;
  [key: string]: any;
}

interface EcosystemStats {
  chains: Array<{ id: number; [key: string]: any }>;
  serviceTypes: unknown[];
  capabilities: unknown[];
}

export class StatsCollector {
  private apiUrl: string;

  constructor(apiUrl: string) {
    this.apiUrl = apiUrl.replace(/\/$/, '');
  }

  private async fetch(path: string, params?: Record<string, string>): Promise<any> {
    try {
      const url = new URL(path, this.apiUrl);
      if (params) {
        Object.entries(params).forEach(([key, value]) => {
          url.searchParams.append(key, value);
        });
      }

      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), 15000);

      try {
        const response = await fetch(url.toString(), {
          signal: controller.signal,
        });

        if (!response.ok) {
          console.error(`[ERROR] API request failed ${path}: ${response.status}`);
          return null;
        }

        return response.json();
      } finally {
        clearTimeout(timeoutId);
      }
    } catch (err) {
      console.error(`[ERROR] API request failed ${path}:`, err);
      return null;
    }
  }

  async fetchEcosystemStats(): Promise<EcosystemStats | null> {
    return this.fetch('/api/agents/filter-options');
  }

  async fetchTrendingAgents(): Promise<Agent[] | null> {
    const data = await this.fetch('/api/agents/suggestions');
    if (Array.isArray(data)) {
      return data;
    }
    return data?.data || null;
  }

  async fetchTopAgents(count: number = 5): Promise<Agent[] | null> {
    const data = await this.fetch('/api/agents', {
      sort: 'score',
      pageSize: String(count),
    });
    if (Array.isArray(data)) {
      return data;
    }
    return data?.data || null;
  }

  async fetchAgentDetail(chainId: number, agentId: string): Promise<any> {
    return this.fetch(`/api/agents/${chainId}/${agentId}`);
  }

  async collectForHighlight(): Promise<any> {
    const trending = await this.fetchTrendingAgents();
    if (!trending || trending.length === 0) {
      return null;
    }

    const agent = trending[0];
    const chainId = (agent.chain_id || agent.chainId) as number | undefined;
    const agentId = (agent.agent_id || agent.agentId) as string | undefined;

    if (!chainId || !agentId) {
      return null;
    }

    const detail = await this.fetchAgentDetail(chainId, agentId);
    if (!detail) {
      return null;
    }

    detail.chain_name = CHAIN_NAMES[chainId] || `Chain ${chainId}`;
    const slug = CHAIN_SLUGS[chainId] || String(chainId);
    detail.url = `${this.apiUrl}/agents/${slug}/${agentId}`;

    return detail;
  }

  async collectForAnomaly(): Promise<any> {
    const top = await this.fetchTopAgents(10);
    const trending = await this.fetchTrendingAgents();

    if (!top && !trending) {
      return null;
    }

    const agents = top || trending || [];
    for (const a of agents) {
      const chainId = a.chain_id || a.chainId;
      const agentId = a.agent_id || a.agentId;

      if (chainId && agentId) {
        const slug = CHAIN_SLUGS[chainId] || String(chainId);
        a.chain_name = CHAIN_NAMES[chainId] || `Chain ${chainId}`;
        a.url = `${this.apiUrl}/agents/${slug}/${agentId}`;
      }
    }

    return { top_agents: top || [], trending: trending || [] };
  }

  async collectForComparison(): Promise<any> {
    const ecosystem = await this.fetchEcosystemStats();
    const top = await this.fetchTopAgents(10);

    if (!ecosystem) {
      return null;
    }

    const chains = ecosystem.chains || [];
    for (const c of chains) {
      c.name = CHAIN_NAMES[c.id] || `Chain ${c.id}`;
    }

    return {
      chains,
      serviceTypes: ecosystem.serviceTypes || [],
      capabilities: ecosystem.capabilities || [],
      top_agents: top || [],
    };
  }
}

interface TweeterResult {
  tweets_generated: number;
  skipped_reason?: string;
  tweet_type?: string;
}

export class Tweeter {
  private statsCollector: StatsCollector;
  private client: Anthropic;
  private readonly DAILY_TWEET_CAP = config.DAILY_TWEET_CAP;

  constructor(statsCollector: StatsCollector) {
    this.statsCollector = statsCollector;
    this.client = new Anthropic({
      apiKey: config.anthropic_api_key,
    });
  }

  private async getTodayTweetCount(): Promise<number> {
    const now = new Date();
    const today = `${now.getUTCFullYear()}-${String(now.getUTCMonth() + 1).padStart(2, '0')}-${String(now.getUTCDate()).padStart(2, '0')}T00:00:00+00:00`;

    // Simple in-memory count (simplified version)
    return 0;
  }

  private async getNextTweetType(): Promise<string> {
    // Get all original tweets from today, count by type
    // Then return the type with the fewest tweets (or the next in cycle)
    const now = new Date();
    const today = `${now.getUTCFullYear()}-${String(now.getUTCMonth() + 1).padStart(2, '0')}-${String(now.getUTCDate()).padStart(2, '0')}T00:00:00+00:00`;

    // For simplicity, just cycle through types in order
    // Get today's tweets and find which type was last
    const count = await this.getTodayTweetCount();
    const tweetTypeIndex = count % TWEET_TYPES.length;
    return TWEET_TYPES[tweetTypeIndex];
  }

  private async collectData(tweetType: string): Promise<any> {
    if (tweetType === 'agent_highlight') {
      return this.statsCollector.collectForHighlight();
    } else if (tweetType === 'anomaly') {
      return this.statsCollector.collectForAnomaly();
    } else if (tweetType === 'comparison') {
      return this.statsCollector.collectForComparison();
    }
    return null;
  }

  private async generateTweet(tweetType: string, data: any): Promise<string | null> {
    const prompt = PROMPTS[tweetType].replace(
      '{data_json}',
      JSON.stringify(data, null, 2)
    );

    let retries = 0;
    const maxRetries = 3;

    while (retries < maxRetries) {
      try {
        const message = await this.client.messages.create({
          model: 'claude-haiku-4-5-20251001',
          max_tokens: 200,
          system: SYSTEM_PROMPT,
          messages: [
            {
              role: 'user',
              content: prompt,
            },
          ],
        });

        let text = message.content[0]?.type === 'text' ? message.content[0].text : null;

        if (!text) {
          retries++;
          continue;
        }

        // Check 160 character limit
        if (text.length > 160) {
          retries++;
          console.log(
            `[INFO] Generated tweet exceeds 160 chars (${text.length}), retrying (${retries}/${maxRetries})...`
          );
          continue;
        }

        return text;
      } catch (err) {
        console.error('[ERROR] Tweet generation failed:', err);
        retries++;
      }
    }

    console.error('[ERROR] Failed to generate valid tweet after retries');
    return null;
  }

  async run(): Promise<TweeterResult> {
    const result: TweeterResult = { tweets_generated: 0 };

    const todayCount = await this.getTodayTweetCount();
    if (todayCount >= this.DAILY_TWEET_CAP) {
      result.skipped_reason = 'daily_tweet_cap_reached';
      return result;
    }

    const tweetType = await this.getNextTweetType();
    const data = await this.collectData(tweetType);

    if (!data) {
      result.skipped_reason = 'no_data_available';
      return result;
    }

    const tweetText = await this.generateTweet(tweetType, data);

    if (!tweetText) {
      result.skipped_reason = 'generation_failed';
      return result;
    }

    const tweetId = uuidv4();
    await db.saveReply({
      id: tweetId,
      tweet_id: tweetId,
      draft_text: tweetText,
      final_text: null,
      status: 'notified',
      source_type: 'original_tweet',
      reviewed_at: null,
      posted_at: null,
      created_at: new Date().toISOString(),
    });

    await telegram.sendResult(
      {
        tweet_id: tweetId,
        content: `[Original Tweet — ${tweetType}]`,
        author_username: 'agent_lynx',
        author_bio: '',
        thread_context: null,
        source_type: 'keyword',
        source_value: 'system',
        created_at: new Date().toISOString(),
        fetched_at: new Date().toISOString(),
        metrics: {},
      },
      tweetText,
      '📌 Original Tweet'
    );

    result.tweets_generated = 1;
    result.tweet_type = tweetType;
    return result;
  }
}
